src/storage.cc

```diff
@@ -0,0 +1,219 @@
+
+  if( ready_time() >= 0 || residence_time_() == 0  && !inventory.empty() ) {
+    ReadyMatl_(ready_time());  // place processing into ready
+  }
+
+  ProcessMat_(throughput_()); // place ready into stocks
+
+  LOG(cyclus::LEV_INFO3, "ComCnv") << "}";
+}
+
+//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+void Storage::AddMat_(cyclus::Material::Ptr mat) {
+
+  LOG(cyclus::LEV_INFO5, "ComCnv") << prototype() << " is initially holding "
+                                << inventory.quantity() << " total.";
+
+  try {
+    inventory.Push(mat);
+  } catch (cyclus::Error& e) {
+    e.msg(Agent::InformErrorMsg(e.msg()));
+    throw e;
+  }
+
+  LOG(cyclus::LEV_INFO5, "ComCnv") << prototype() << " added " << mat->quantity()
+                                << " of material to its inventory, which is holding "
+                                << inventory.quantity() << " total.";
+
+}
+
+//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+void Storage::BeginProcessing_(){
+  while( inventory.count() > 0 ){
+    try {
+      processing.Push(inventory.Pop());
+      entry_times.push_back(context()->time());
+
+      LOG(cyclus::LEV_DEBUG2, "ComCnv") << "Storage " << prototype() 
+                                      << " added resources to processing at t= "
+                                      << context()->time();
+    } catch (cyclus::Error& e) {
+      e.msg(Agent::InformErrorMsg(e.msg()));
+      throw e;
+    }
+  }
+}
+
+//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+void Storage::ProcessMat_(double cap){
+  using cyclus::Material;
+  using cyclus::ResCast;
+  using cyclus::toolkit::ResBuf;
+  using cyclus::toolkit::Manifest;
+
+  if ( !ready.empty() ){
+    try { 
+
+      double max_pop = std::min(cap, ready.quantity());
+
+      if ( batch_handling_() ) {
+        if (max_pop == ready.quantity()) {
+          stocks.Push(ready.PopN(ready.count()));
+        }
+        else {
+          double cap_pop = ready.Peek()->quantity();
+          while(cap_pop<=max_pop && !ready.empty() ){
+            stocks.Push(ready.Pop());
+            cap_pop += ready.empty() ? 0 : ready.Peek()->quantity();
+          }
+        }
+      }
+      else {
+        stocks.Push(ready.Pop(max_pop));
+      }
+
+      LOG(cyclus::LEV_INFO1, "ComCnv") << "Storage " << prototype() 
+                                        << " moved resources" 
+                                        << " from ready to stocks" 
+                                        << " at t= " << context()->time();
+    } catch (cyclus::Error& e) {
+      e.msg(Agent::InformErrorMsg(e.msg()));
+      throw e;
+    }
+  }
+}
+
+//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+void Storage::ReadyMatl_(int time) {
+  using cyclus::toolkit::ResBuf;
+
+  int to_ready = 0;
+
+  while(!entry_times.empty() && entry_times.front()<=time){
```

Comment:
(non-blocking comment)

While I think the algorithm is correct, I find the grammar of this algorithm to be a little confusing.

You could track "release_times" instead of "entry_times" and then just compare to current times.

Alternatively, you could use a different name for the argument "time", although I'm not sure there is a grammatically straightforward choice.